pennylane/beta/tapes/cv_param_shift.py

```diff
@@ -0,0 +1,458 @@
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        recipe = op.grad_recipe[p_idx]
+        c, s = (0.5, np.pi / 2) if recipe is None else recipe
+
+        shift = np.zeros_like(params)
+        shift[idx] = s
+
+        shift_forward = np.array(self.execute_device(params + shift, device))
+        shift_backward = np.array(self.execute_device(params - shift, device))
+
+        return c * (shift_forward - shift_backward)
+
+    def parameter_shift_second_order(self, idx, device, params, **options):
+        r"""Partial derivative using the second-order CV parameter-shift rule of a
+        tape consisting of *only* expectation values of observables.
+
+        This method supports the gradient of expectation values of first-
+        and second-order observables.
+
+        .. warning::
+
+            This method can only be executed on devices that support the
+            :class:`~.PolyXP` observable.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        # pylint: disable=protected-access
+
+        if "PolyXP" not in device.observables:
+            # If the device does not support PolyXP, must fallback
+            # to numeric differentiation.
+            warnings.warn(
+                f"The device {device.short_name} does not support "
+                "the PolyXP observable. The analytic parameter-shift cannot be used for "
+                "second-order observables; falling back to finite-differences.",
+                UserWarning,
+            )
+            return self.numeric_pd(idx, device, params, **options)
+
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        recipe = op.grad_recipe[p_idx]
+        c, s = (0.5, np.pi / 2) if recipe is None else recipe
+
+        shift = np.zeros_like(params)
+        shift[idx] = s
+
+        # evaluate transformed observables at the original parameter point
+        # first build the Heisenberg picture transformation matrix Z
+        self.set_parameters(params + shift)
+        Z2 = op.heisenberg_tr(device.wires)

```

Comment:
Seems to be the same for `CVQNode` too, though got me thinking: shouldn't this use the wires of the observable  ---- `Z2 = op.heisenberg_tr(op.wires)`? :thinking: 

It seems to be fine in terms of the result, but I think we're creating matrices according to the entire system defined for the device:
```python
from pennylane.wires import Wires

dev_wires = Wires([0,1,2,3])

qml.Rotation(0.4, wires=[0]).heisenberg_tr(dev_wires).shape
```
Outputs:
```
(9, 9)
```
Whereas:
```python
from pennylane.wires import Wires

dev_wires = Wires([0,1]) # <--- only 2 wires

qml.Rotation(0.4, wires=[0]).heisenberg_tr(dev_wires).shape
```
Outputs:
```
(5, 5)
```
Perhaps this doesn't affect the observable transformation as `Z` is later conjugated? :thinking:

Reply:
Hey Antal, we need to work on the entire device wires here.

To see why, consider the case

```python
@qml.qnode(dev)
def circuit(x):
    qml.Rotation(x, wires=0)
    qml.Beamsplitter(0.5, 0.1, wires=[0, 1])
    return qml.expval(qml.NumberOperator(0))
```

Even though the gate we are differentiating and the corresponding observable only apply to wire 0, the Gaussian transformation inbetween --- the beamsplitter --- applies to _two_ wires. So `Z` will be size `[4, 4]`, but if we were to only do `obs.heisenberg_tr(obs.wires)`, we would end up with an `A` that was `[2,2]`!

Expanding the heisenberg transformation on the device wires ensures that A and Z will always be the same shape, allowing us to do the matrix multiplication.

Reply:
That's a fair point, thanks! Think the `wires` argument in the docstrings for `heisenberg_expand`, `heisenberg_tr` and `heisenberg_obs` might need to be updated:
https://github.com/PennyLaneAI/pennylane/blob/5b4aa7b3b02796b87545d32d26f06cf5674d08c4/pennylane/operation.py#L1415

Another less frequent case that I think would still remain is when there are unused wires defined for the device.

Reply:
Yep good point, we should update that docstring!

Reply:
Updated